File: __Viejo/funciones/cobros.vb
Imports System.Data.SqlClient

Module cobros
    ' ************************************ FUNCIONES DE COBROS ***************************
    Public Function info_cobro(ByVal id_cobro As String) As cobro
        Dim tmp As New cobro
        Dim sqlstr As String

        sqlstr = "SET DATEFORMAT dmy; SELECT id_cobro, ISNULL(id_cobro_oficial, -1), ISNULL(id_cobro_no_oficial, -1), fecha_carga, fecha_cobro, id_cliente, id_cc, dineroEnCc, efectivo, totalTransferencia, totalCh, " &
                    "totalRetencion, total, hayCheque, hayTransferencia, hayRetencion, activo, ISNULL(id_anulaCobro, -1), notas, firmante " &
                    "FROM cobros " &
                    "WHERE id_cobro = '" + id_cobro + "'"

        Try
            'Crea y abre una nueva conexión
            abrirdb(serversql, basedb, usuariodb, passdb)

            'Propiedades del SqlCommand
            Dim comando As New SqlCommand
            With comando
                .CommandType = CommandType.Text
                .CommandText = sqlstr
                .Connection = CN
            End With

            Dim da As New SqlDataAdapter 'Crear nuevo SqlDataAdapter
            Dim dataset As New DataSet 'Crear nuevo dataset

            da.SelectCommand = comando

            'llenar el dataset
            da.Fill(dataset, "Tabla")
            tmp.id_cobro = dataset.Tables("tabla").Rows(0).Item(0).ToString
            tmp.id_cobro_oficial = dataset.Tables("tabla").Rows(0).Item(1).ToString
            tmp.id_cobro_no_oficial = dataset.Tables("tabla").Rows(0).Item(2).ToString
            tmp.fecha_carga = dataset.Tables("tabla").Rows(0).Item(3).ToString
            tmp.fecha_cobro = dataset.Tables("tabla").Rows(0).Item(4).ToString
            tmp.id_cliente = dataset.Tables("tabla").Rows(0).Item(5).ToString
            tmp.id_cc = dataset.Tables("tabla").Rows(0).Item(6).ToString
            tmp.dineroEnCc = dataset.Tables("tabla").Rows(0).Item(7).ToString
            tmp.efectivo = dataset.Tables("tabla").Rows(0).Item(8).ToString
            tmp.totalTransferencia = dataset.Tables("tabla").Rows(0).Item(9).ToString
            tmp.totalCh = dataset.Tables("tabla").Rows(0).Item(10).ToString
            tmp.totalRetencion = dataset.Tables("tabla").Rows(0).Item(11).ToString
            tmp.total = dataset.Tables("tabla").Rows(0).Item(12).ToString
            tmp.hayCheque = dataset.Tables("tabla").Rows(0).Item(13).ToString
            tmp.hayTransferencia = dataset.Tables("tabla").Rows(0).Item(14).ToString
            tmp.hayRetencion = dataset.Tables("tabla").Rows(0).Item(15).ToString
            tmp.activo = dataset.Tables("tabla").Rows(0).Item(16).ToString
            tmp.id_anulaCobro = dataset.Tables("tabla").Rows(0).Item(17).ToString
            tmp.notas = dataset.Tables("tabla").Rows(0).Item(18).ToString
            tmp.firmante = dataset.Tables("tabla").Rows(0).Item(19).ToString
            cerrardb()
            Return tmp
        Catch ex As Exception
            MsgBox(ex.Message.ToString)
            tmp.efectivo = -1
            cerrardb()
            Return tmp
        End Try
    End Function

    Public Function addcobro(ByVal c As cobro) As Integer
        Dim sqlComm As New SqlCommand 'Comando en el resto
        Dim resultado As Integer

        Try
            abrirdb(serversql, basedb, usuariodb, passdb)

            With sqlComm
                .CommandText = "SP_insertCobro"
                .CommandType = CommandType.StoredProcedure

                With .Parameters
                    .AddWithValue("id_cobro_oficial", c.id_cobro_oficial)
                    .AddWithValue("id_cobro_no_oficial", c.id_cobro_no_oficial)
                    .AddWithValue("fecha_cobro", c.fecha_cobro)
                    .AddWithValue("id_cliente", c.id_cliente)
                    .AddWithValue("id_cc", c.id_cc)
                    .AddWithValue("dineroEnCc", c.dineroEnCc)
                    .AddWithValue("efectivo", c.efectivo)
                    .AddWithValue("totalTransferencia", c.totalTransferencia)
                    .AddWithValue("totalCh", c.totalCh)
                    .AddWithValue("totalRetencion", c.totalRetencion)
                    .AddWithValue("total", c.total)
                    .AddWithValue("hayCheque", c.hayCheque)
                    .AddWithValue("hayTransferencia", c.hayTransferencia)
                    .AddWithValue("hayRetencion", c.hayRetencion)
                    .AddWithValue("id_anulaCobro", c.id_anulaCobro)
                    .AddWithValue("notas", c.notas)
                    .AddWithValue("firmante", c.firmante)
                    .Add(New SqlParameter("@resultado", SqlDbType.Int)).Direction = ParameterDirection.Output
                End With
                .Connection = CN
                .ExecuteNonQuery()
            End With

            resultado = CInt(sqlComm.Parameters("@resultado").Value)
            Return resultado
        Catch ex As Exception
            MsgBox(ex.Message)
            Return False
        Finally
            cerrardb()
        End Try
    End Function

    Public Function borrarcobro(ByVal c As cobro) As Boolean
        abrirdb(serversql, basedb, usuariodb, passdb)
        Dim ac As New cobro
        Dim cc As New ccCliente

        ac = c
        With ac
            .dineroEnCc = .dineroEnCc * -1
            .efectivo = .efectivo * -1
            .totalTransferencia = .totalTransferencia * -1
            .totalCh = .totalCh * -1
            .totalRetencion = .totalRetencion * -1
            .total = .total * -1
            .id_anulaCobro = c.id_cobro
            If .id_cobro_no_oficial = -1 Then
                .notas += vbCr + "Anula cobro: " + .id_cobro_oficial.ToString
            Else
                .notas += vbCr + "Anula cobro: " + .id_cobro_no_oficial.ToString
            End If
        End With

        'Agrego un cobro exactamente al revez, referenciando al cobro original
        ac.id_cobro = addcobro(ac)

        'Borro todas las transferencias que pertenecen a ese cobro
        borrar_transferencias(c.id_cobro)

        'Borro todas las retenciones que pertenecen a ese cobro
        borrar_retenciones(c.id_cobro)

        If ac.id_cobro Then
            'Borro los cheques asignados al cobro para liberarlos y actualizo el saldo
            borrar_chequeCobrado(ac.id_cobro)
            cc = info_ccCliente(c.id_cc)
            cc.saldo -= c.total
            updateCCCliente(cc)
            Return True
        Else
            Return False
        End If
    End Function

    Public Function add_chequeCobrado(ByVal id_cobro As Integer, ByVal id_cheque() As Integer) As Integer
        abrirdb(serversql, basedb, usuariodb, passdb)

        Dim mytrans As SqlTransaction
        Dim Comando As New SqlClient.SqlCommand
        Dim sqlstr As String = ""

        mytrans = CN.BeginTransaction()

        Try

            For Each cheque As Integer In id_cheque
                sqlstr += "INSERT INTO cobros_cheques (id_cobro, id_cheque) VALUES (" + id_cobro.ToString + ", " + cheque.ToString + "); "
            Next

            sqlstr = Left(Trim(sqlstr), (Len(Trim(sqlstr)) - 1))

            Comando = New SqlClient.SqlCommand(sqlstr, CN)

            Comando.Transaction = mytrans
            Comando.ExecuteNonQuery()

            mytrans.Commit()
            cerrardb()
            Return True
        Catch ex As Exception
            MsgBox(ex.Message)
            cerrardb()
            Return False
        End Try
    End Function

    Public Function borrar_chequeCobrado(ByVal id_cobro As Integer) As Boolean
        abrirdb(serversql, basedb, usuariodb, passdb)

        Dim mytrans As SqlTransaction
        Dim Comando As New SqlClient.SqlCommand
        Dim sqlstr As String

        mytrans = CN.BeginTransaction()

        Try
            sqlstr = "DELETE FROM cobros_cheques WHERE id_cobro = '" + id_cobro.ToString + "'"
            Comando = New SqlClient.SqlCommand(sqlstr, CN)
            Comando.Transaction = mytrans
            Comando.ExecuteNonQuery()

            mytrans.Commit()
            cerrardb()
            Return True
        Catch ex As Exception
            MsgBox(ex.Message.ToString)
            cerrardb()
            Return False
        End Try
    End Function

    Public Function liberar_chequesCobrados(ByVal id_cobro As Integer) As Boolean
        abrirdb(serversql, basedb, usuariodb, passdb)

        Dim mytrans As SqlTransaction
        Dim Comando As New SqlClient.SqlCommand
        Dim sqlstr As String

        mytrans = CN.BeginTransaction()

        Try
            sqlstr = "UPDATE ch " &
                    "SET ch.id_estadoch = '" + ID_CH_CARTERA + "' " &
                    "FROM cheques AS ch " &
                    "INNER JOIN cobros_cheques AS cch ON ch.id_cheque = cch.id_cheque " &
                    "WHERE cch.id_cobro = '" + id_cobro.ToString + "' "
            Comando = New SqlClient.SqlCommand(sqlstr, CN)
            Comando.Transaction = mytrans
            Comando.ExecuteNonQuery()

            mytrans.Commit()
            cerrardb()
            Return True
        Catch ex As Exception
            MsgBox(ex.Message.ToString)
            cerrardb()
            Return False
        End Try
    End Function

    Public Function borrar_transferencias(ByVal id_cobro As Integer) As Boolean
        abrirdb(serversql, basedb, usuariodb, passdb)

        Dim mytrans As SqlTransaction
        Dim Comando As New SqlClient.SqlCommand
        Dim sqlstr As String

        mytrans = CN.BeginTransaction()

        Try
            sqlstr = "DELETE FROM transferencias WHERE id_cobro = '" + id_cobro.ToString + "'"
            Comando = New SqlClient.SqlCommand(sqlstr, CN)
            Comando.Transaction = mytrans
            Comando.ExecuteNonQuery()

            mytrans.Commit()
            cerrardb()
            Return True
        Catch ex As Exception
            MsgBox(ex.Message.ToString)
            cerrardb()
            Return False
        End Try
    End Function

    Public Function borrar_retenciones(ByVal id_cobro As Integer) As Boolean
        abrirdb(serversql, basedb, usuariodb, passdb)

        Dim mytrans As SqlTransaction
        Dim Comando As New SqlClient.SqlCommand
        Dim sqlstr As String

        mytrans = CN.BeginTransaction()

        Try
            sqlstr = "DELETE FROM cobros_retenciones WHERE id_cobro = '" + id_cobro.ToString + "'"
            Comando = New SqlClient.SqlCommand(sqlstr, CN)
            Comando.Transaction = mytrans
            Comando.ExecuteNonQuery()

            mytrans.Commit()
            cerrardb()
            Return True
        Catch ex As Exception
            MsgBox(ex.Message.ToString)
            cerrardb()
            Return False
        End Try
    End Function

    Public Function guardar_cobros_facturas_importes(ByVal id_cobro As Integer, ByVal dg As DataGridView) As Boolean
        abrirdb(serversql, basedb, usuariodb, passdb)

        Dim mytrans As SqlTransaction
        Dim Comando As New SqlClient.SqlCommand
        Dim sqlstr As String = "SET NOCOUNT ON; SET DATEFORMAT dmy; "

        mytrans = CN.BeginTransaction()

        Try
            If dg.Rows.Count > 0 Then
                For Each row As DataGridViewRow In dg.Rows
                    If Not row Is Nothing And Not row.IsNewRow Then
                        sqlstr += "INSERT INTO cobros_Nfacturas_importes (id_cobro, fecha, nfactura, importe) VALUES ('" + id_cobro.ToString + "', '" + row.Cells("Fecha").Value.ToString + "', '" + row.Cells("Factura").Value.ToString + "', '" + row.Cells("Importe").Value.ToString + "'); "
                    End If
                Next

                sqlstr = Left(Trim(sqlstr), (Len(Trim(sqlstr)) - 1))

                Comando = New SqlClient.SqlCommand(sqlstr, CN)

                Comando.Transaction = mytrans
                Comando.ExecuteNonQuery()

                mytrans.Commit()
                Return True
            Else
                Return False
            End If
        Catch ex As Exception
            MsgBox(ex.Message)
            Return False
        Finally
            cerrardb()
        End Try

    End Function


    Public Function Cobro_Ya_Anulado(ByVal id_cobro As String) As Boolean
        Dim tmp As Integer = -1
        Dim sqlstr As String

        sqlstr = "SELECT ISNULL(id_anulaCobro, -1) " &
                    "FROM cobros " &
                    "WHERE id_anulaCobro = '" + id_cobro + "'"

        Try
            'Crea y abre una nueva conexión
            abrirdb(serversql, basedb, usuariodb, passdb)

            'Propiedades del SqlCommand
            Dim comando As New SqlCommand
            With comando
                .CommandType = CommandType.Text
                .CommandText = sqlstr
                .Connection = CN
            End With

            Dim da As New SqlDataAdapter 'Crear nuevo SqlDataAdapter
            Dim dataset As New DataSet 'Crear nuevo dataset

            da.SelectCommand = comando

            'llenar el dataset
            da.Fill(dataset, "Tabla")
            If dataset.Tables("Tabla").Rows.Count > 0 Then
                tmp = dataset.Tables("tabla").Rows(0).Item(0).ToString
            End If

            If tmp <> -1 Then
                Return True
            Else
                Return False
            End If
        Catch ex As Exception
            MsgBox(ex.Message.ToString)
            Return False
        Finally
            cerrardb()
        End Try
    End Function

    Public Function Ultimo_cobro_oficial() As Integer
        Dim id_cobro_oficial As Integer
        Dim sqlstr As String

        sqlstr = "SELECT MAX(id_cobro_oficial) FROM cobros"

        Try
            'Crea y abre una nueva conexión
            abrirdb(serversql, basedb, usuariodb, passdb)

            'Propiedades del SqlCommand
            Dim comando As New SqlCommand
            With comando
                .CommandType = CommandType.Text
                .CommandText = sqlstr
                .Connection = CN
            End With

            Dim da As New SqlDataAdapter 'Crear nuevo SqlDataAdapter
            Dim dataset As New DataSet 'Crear nuevo dataset

            da.SelectCommand = comando

            'llenar el dataset
            da.Fill(dataset, "Tabla")
            id_cobro_oficial = dataset.Tables("tabla").Rows(0).Item(0)
            If id_cobro_oficial = -1 Then
                id_cobro_oficial = 1
            Else
                id_cobro_oficial += 1
            End If
            Return id_cobro_oficial
        Catch ex As Exception
            MsgBox(ex.Message.ToString)
            Return -1
        Finally
            cerrardb()
        End Try
    End Function

    Public Function Ultimo_cobro_no_oficial() As Integer
        Dim id_cobro_no_oficial As Integer
        Dim sqlstr As String

        sqlstr = "SELECT MAX(id_cobro_no_oficial) FROM cobros"

        Try
            'Crea y abre una nueva conexión
            abrirdb(serversql, basedb, usuariodb, passdb)

            'Propiedades del SqlCommand
            Dim comando As New SqlCommand
            With comando
                .CommandType = CommandType.Text
                .CommandText = sqlstr
                .Connection = CN
            End With

            Dim da As New SqlDataAdapter 'Crear nuevo SqlDataAdapter
            Dim dataset As New DataSet 'Crear nuevo dataset

            da.SelectCommand = comando

            'llenar el dataset
            da.Fill(dataset, "Tabla")
            id_cobro_no_oficial = dataset.Tables("tabla").Rows(0).Item(0)
            If id_cobro_no_oficial = -1 Then
                id_cobro_no_oficial = 1
            Else
                id_cobro_no_oficial += 1
            End If
            Return id_cobro_no_oficial
        Catch ex As Exception
            MsgBox(ex.Message.ToString)
            Return -1
        Finally
            cerrardb()
        End Try
    End Function
    ' ************************************ FUNCIONES DE COBROS ***************************
End Module

File: funciones/cobros.cs
using System;
using System.Data;
using System.Linq;
using System.Windows.Forms;

namespace Centrex
{

    static class cobros
    {
        // ************************************ FUNCIONES DE COBROS ***************************
        public static CobroEntity info_cobro(string id_cobro)
        {
            try
            {
                using (var context = new CentrexDbContext())
                {
                    return context.CobroEntity.FirstOrDefault(c => c.IdCobro == Conversions.ToInteger(id_cobro));
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox(ex.Message.ToString());
                return null;
            }
        }

        public static int addcobro(cobro c)
        {
            try
            {
                using (var context = new CentrexDbContext())
                {
                    var cobroEntity = new CobroEntity()
                    {
                        IdCobroOficial = c.id_cobro_oficial,
                        IdCobroNoOficial = c.id_cobro_no_oficial,
                        FechaCarga = DateOnly.FromDateTime(DateTime.Now),
                        FechaCobro = !string.IsNullOrEmpty(c.fecha_cobro) ? DateOnly.FromDateTime(Conversions.ToDate(c.fecha_cobro)) : DateOnly.FromDateTime(DateTime.Now),
                        IdCliente = c.id_cliente,
                        IdCc = c.id_cc,
                        DineroEnCc = (decimal)c.dineroEnCc,
                        Efectivo = (decimal)c.efectivo,
                        TotalTransferencia = (decimal)c.totalTransferencia,
                        TotalCh = (decimal)c.totalCh,
                        TotalRetencion = (decimal)c.totalRetencion,
                        Total = (decimal)c.total,
                        HayCheque = c.hayCheque,
                        HayTransferencia = c.hayTransferencia,
                        HayRetencion = c.hayRetencion,
                        Activo = true,
                        IdAnulaCobro = c.id_anulaCobro != -1 ? c.id_anulaCobro : default,
                        Notas = c.notas ?? "",
                        Firmante = c.firmante ?? ""
                    };

                    context.CobroEntity.Add(cobroEntity);
                    context.SaveChanges();

                    return cobroEntity.IdCobro;
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox(ex.Message);
                return -1;
            }
        }

        public static bool borrarcobro(cobro c)
        {
            try
            {
                var ac = new cobro();
                var cc = new ccCliente();

                ac = c;
                ac.dineroEnCc = ac.dineroEnCc * -1;
                ac.efectivo = ac.efectivo * -1;
                ac.totalTransferencia = ac.totalTransferencia * -1;
                ac.totalCh = ac.totalCh * -1;
                ac.totalRetencion = ac.totalRetencion * -1;
                ac.total = ac.total * -1;
                ac.id_anulaCobro = c.id_cobro;
                if (ac.id_cobro_no_oficial == -1)
                {
                    ac.notas += Constants.vbCr + "Anula cobro: " + ac.id_cobro_oficial.ToString();
                }
                else
                {
                    ac.notas += Constants.vbCr + "Anula cobro: " + ac.id_cobro_no_oficial.ToString();
                }

                // Agrego un cobro exactamente al revez, referenciando al cobro original
                ac.id_cobro = addcobro(ac);

                // Borro todas las transferencias que pertenecen a ese cobro
                borrar_transferencias(c.id_cobro);

                // Borro todas las retenciones que pertenecen a ese cobro
                borrar_retenciones(c.id_cobro);

                if (ac.id_cobro > 0)
                {
                    // Borro los cheques asignados al cobro para liberarlos y actualizo el saldo
                    borrar_chequeCobrado(ac.id_cobro);
                    cc = ccClientes.info_ccCliente(c.id_cc);
                    cc.saldo -= c.total;
                    ccClientes.updateCCCliente(cc);
                    return true;
                }
                else
                {
                    return false;
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox(ex.Message);
                return false;
            }
        }

        public static bool add_chequeCobrado(int id_cobro, int[] cheques)
        {
            try
            {
                using (var context = new CentrexDbContext())
                {
                    foreach (var idCheque in cheques)
                    {
                        // Buscar el cheque y actualizar estado
                        var ch = context.ChequeEntity.FirstOrDefault(c => c.IdCheque == idCheque);
                        if (ch is not null)
                        {
                            ch.IdEstadoch = VariablesGlobales.ID_CH_ENTREGADO;   // Constante definida en mconfig
                            ch.FechaSalida = DateOnly.FromDateTime(DateTime.Now);
                            context.Entry(ch).State = EntityState.Modified;
                        }

                        // Crear la relación Cobro-Cheque
                        var cobroCheque = new CobroChequeEntity()
                        {
                            IdCobro = id_cobro,
                            IdCheque = idCheque
                        };
                        context.CobroChequeEntity.Add(cobroCheque);
                    }

                    context.SaveChanges();
                    return true;
                }
            }

            catch (Exception ex)
            {
                Interaction.MsgBox("Error al vincular cheques con el cobro: " + ex.Message, MsgBoxStyle.Critical, "Centrex");
                return false;
            }
        }


        public static bool borrar_chequeCobrado(int id_cobro)
        {
            try
            {
                using (var context = new CentrexDbContext())
                {
                    var chequesCobrados = context.CobroChequeEntity.Where(cc => cc.IdCobro == id_cobro).ToList();

                    foreach (var cobroCheque in chequesCobrados)
                        context.CobroChequeEntity.Remove((CobroChequeEntity)cobroCheque);

                    context.SaveChanges();
                    return true;
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox(ex.Message.ToString());
                return false;
            }
        }

        public static bool liberar_chequesCobrados(int id_cobro)
        {
            try
            {
                using (var context = new CentrexDbContext())
                {
                    // Obtener los IDs de cheques asociados al cobro
                    var idsChequesQuery = from cc in context.CobroChequeEntity
                                          where cc.IdCobro == id_cobro
                                          select cc.IdCheque;

                    var idsCheques = idsChequesQuery.ToList();

                    // Actualizar el estado de los cheques a CARTERA
                    foreach (var idCheque in idsCheques)
                    {
                        var cheque = context.ChequeEntity.FirstOrDefault(ch => ch.IdCheque == idCheque);
                        if (cheque is not null)
                        {
                            cheque.IdEstadoch = VariablesGlobales.ID_CH_CARTERA;
                        }
                    }

                    context.SaveChanges();
                    return true;
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox(ex.Message.ToString());
                return false;
            }
        }

        public static bool borrar_transferencias(int id_cobro)
        {
            try
            {
                using (var context = new CentrexDbContext())
                {
                    var transferencias = context.TransferenciaEntity.Where(t => t.IdCobro == id_cobro).ToList();

                    foreach (var transferencia in transferencias)
                        context.TransferenciaEntity.Remove(transferencia);

                    context.SaveChanges();
                    return true;
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox(ex.Message.ToString());
                return false;
            }
        }

        public static bool borrar_retenciones(int id_cobro)
        {
            try
            {
                using (var context = new CentrexDbContext())
                {
                    var retenciones = context.CobroRetencionEntity.Where(r => r.IdCobro == id_cobro).ToList();

                    foreach (var retencion in retenciones)
                        context.CobroRetencionEntity.Remove((CobroRetencionEntity)retencion);

                    context.SaveChanges();
                    return true;
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox(ex.Message.ToString());
                return false;
            }
        }

        public static bool guardar_cobros_facturas_importes(int id_cobro, DataGridView dg_view)
        {
            try
            {
                using (var context = new CentrexDbContext())
                {
                    // Eliminar relaciones previas del mismo cobro
                    var previas = context.CobroNfacturaImporteEntity.Where(f => f.IdCobro == id_cobro).ToList();
                    foreach (var f in previas)
                        context.CobroNfacturaImporteEntity.Remove((CobroNfacturaImporteEntity)f);

                    // Agregar nuevas relaciones desde el DataGridView
                    foreach (DataGridViewRow row in dg_view.Rows)
                    {
                        if (row.IsNewRow)
                            continue;
                        if (row.Cells["ID"].Value is null)
                            continue;

                        string nFactura = Convert.ToString(row.Cells["ID"].Value);
                        decimal importe = Convert.ToDecimal(row.Cells["Importe"].Value);
                        var fecha = row.Cells["Fecha"] is not null && row.Cells["Fecha"].Value is not null ? Convert.ToDateTime(row.Cells["Fecha"].Value) : DateTime.Now;

                        var nuevaRelacion = new CobroNfacturaImporteEntity()
                        {
                            IdCobro = id_cobro,
                            Nfactura = nFactura,
                            Importe = importe,
                            Fecha = DateOnly.FromDateTime(fecha)
                        };
                        context.CobroNfacturaImporteEntity.Add(nuevaRelacion);
                    }

                    context.SaveChanges();
                    return true;
                }
            }

            catch (Exception ex)
            {
                Interaction.MsgBox("Error al guardar facturas e importes del cobro: " + ex.Message, MsgBoxStyle.Critical, "Centrex");
                return false;
            }
        }

        public static bool Cobro_Ya_Anulado(string id_cobro)
        {
            try
            {
                using (var context = new CentrexDbContext())
                {
                    int idCobroInt = Conversions.ToInteger(id_cobro);
                    bool existe = context.CobroEntity.Any(c => c.IdAnulaCobro.HasValue && c.IdAnulaCobro.Value == idCobroInt);
                    return existe;
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox(ex.Message.ToString());
                return false;
            }
        }

        public static int Ultimo_cobro_oficial()
        {
            try
            {
                using (var context = new CentrexDbContext())
                {
                    if (context.CobroEntity.Any())
                    {
                        var maxIdOficial = context.CobroEntity.Max(c => c.IdCobroOficial);
                        if (maxIdOficial == -1)
                        {
                            return 1;
                        }
                        else
                        {
                            return maxIdOficial + 1;
                        }
                    }
                    else
                    {
                        return 1;
                    }
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox(ex.Message.ToString());
                return -1;
            }
        }

        public static int Ultimo_cobro_no_oficial()
        {
            try
            {
                using (var context = new CentrexDbContext())
                {
                    if (context.CobroEntity.Any())
                    {
                        var maxIdNoOficial = context.CobroEntity.Max(c => c.IdCobroNoOficial);
                        if (maxIdNoOficial == -1)
                        {
                            return 1;
                        }
                        else
                        {
                            return maxIdNoOficial + 1;
                        }
                    }
                    else
                    {
                        return 1;
                    }
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox(ex.Message.ToString());
                return -1;
            }
        }
        // ************************************ FUNCIONES DE COBROS ***************************
    }
}
